Add `menu` option

jarivo opened the pull request:
This PR "fixes" both #14 and #43 

So basically the user can now:
- Specify a ```menu```-property in the configuration to override the default menu structure
- Give a ```transform```-function to the predefined actions in order to manipulate e.g. the data that will be copied to the clipboard
- ~~Override the ```click``` method. This will not work in combination with the transform so if you want to override the default click handler then you're basically on your own.~~

The menuItem itself will determine wheter it should be visible or not. (Although this should be easy to implement when needed)

```append``` and ```prepend``` should still work as expected. Although they now have an extra argument with the default actions.

I've changed the fixture to represent the new features.
```javascript
require('.')({
	labels: {
		cut: 'Configured Cut',
		copy: 'Configured Copy',
		paste: 'Configured Paste',
		save: 'Configured Save Image',
		copyLink: 'Configured Copy Link',
		inspect: 'Configured Inspect'
	},
	prepend: (x) => [x.CUT({transform: (content) => "modified_cut_" + content})],
	menu: (x) => [
		x.SEPARATOR(),
		x.COPY_LINK({transform: (content) => "modified_link_" + content}),
		x.SEPARATOR(),
		{
			label: 'Unicorn'
		},
		x.SEPARATOR(),
		x.COPY({transform: (content) => "modified_copy_" + content}),
		{
			label: 'Invisible',
			visible: false
		},
		x.PASTE({transform: (content) => "modified_paste_" + content})
	],
	append: (x) => [x.SAVE_IMAGE()]
});
```

TODO
- [x] Update README.md
- [x] Fix CI

---

Fixes #14 
Fixes #43

sindresorhus commented on package.json:
```diff
@@ -39,6 +39,13 @@
+          "newIsCap": true,
+          "capIsNew": true,
+          "properties": false
+      }]
```
Why are you adding this?

jarivo replied:
This is removed now, but I've added it because eslint doesn't allow the uppercase function names for the default actions without the new keyword.

sindresorhus commented on fixture.js:
```diff
@@ -10,21 +10,23 @@ require('.')({
+			label: 'Invisible',
+			visible: false
+		},
+		x.PASTE({transform: (content) => "modified_paste_" + content})
```
Looking at this now, it looks pretty ugly with uppercase. Let's just use camelCase.

sindresorhus commented on index.js:
```diff
@@ -15,96 +15,131 @@ function create(win, opts) {
+		}
+
+		if (!opts.menu && props.mediaType === 'image') {
+			menuTpl = [defaultActions.SEPARATOR(), defaultActions.SAVE_IMAGE(), defaultActions.SEPARATOR()];
```
This logic should be in the save image action, not here. Same with the below for "copy link".

sindresorhus commented on fixture.js:
```diff
@@ -10,21 +10,23 @@ require('.')({
-	}],
-	append: () => {}
+	prepend: (x) => [x.CUT({transform: (content) => "modified_cut_" + content})],
+	menu: (x) => [
```
Needs a better name than `x`.

sindresorhus commented:
> Override the `click` method. This will not work in combination with the transform so if you want to override the default click handler then you're basically on your own.

I don't really see the point of this. If you override click, you might as well just create your own menu item.

jarivo commented:
@sindresorhus 

I've also removed this. Initially I added it with the idea of users maybe don't want to deal with "when should a menu item be visible" depending on the context.

jarivo commented:
Any update on this PR?

zeke commented on readme.md:
```diff
@@ -51,17 +51,50 @@ Window or WebView to add the context menu to.
 Type: `Function`
 
-Should return an array of [MenuItem](http://electron.atom.io/docs/api/menu-item/)'s to be prepended to the context menu. The first argument is [this `params` object](http://electron.atom.io/docs/api/web-contents/#event-context-menu). The second argument is the [BrowserWindow](http://electron.atom.io/docs/api/browser-window/) the context menu was requested for.
+Should return an array of [MenuItem](http://electron.atom.io/docs/api/menu-item/)'s to be prepended to the context menu. The first argument is an array of default actions that can be used. The first argument is [this `params` object](http://electron.atom.io/docs/api/web-contents/#event-context-menu). The second argument is the [BrowserWindow](http://electron.atom.io/docs/api/browser-window/) the  ...
```
"The first argument"  appears twice here.

sindresorhus commented:
@jarivo I'm really sorry I never got around to merging this. It just got lost in all my tabs... I'm willing to prioritize reviewing it now if you fix the merge conflict, if not, no worries either.

jarivo commented:
@JoniVR @sindresorhus 

I'm happy to help out fixing the merge conflicts :smile:

JoniVR commented:
@jarivo 
I have some modified files (the conflicting ones) here which resolve the conflicts, but some functionality isn't working 100% yet. Should I send you the contents or just submit a PR with the changes to your fork? If you'd prefer doing everything yourself then that's fine too, it would be great if this PR could get implemented! :D

jarivo commented:
@JoniVR 
Thanks!
I guess the easiest way is a PR to my fork 👍

jarivo commented:
TODO
- [ ] Complete definition file
- [ ] Check README
